core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -282,6 +283,13 @@ public Snapshot apply() {
       throw new RuntimeIOException(e, "Failed to write manifest list file");
     }
 
+    Long addedRows = null;
+    Long lastRowId = null;
+    if (base.rowLineageEnabled()) {
+      addedRows = calculateAddedRows(manifests);
```

Comment:
It's an optimization so we can always just do this later (and arguably makes it a bit harder to read the code) but instead of waiting until all the manifests are written what if we set the addedRows as we added to the writer. 

Something like


```
// remove the writer.addAll
for (ManifestFile manifest: manifestFIles) {
     if (manifest.snapshotId() == null || (manifest.snapshotId() == this.snapshotId) {
                     Preconditions.checkArgument(
                  manifest.addedRowsCount() != null,
                  "Cannot determine number of added rows in snapshot because"
                      + " the entry for manifest %s is missing the field `added-rows-count`",
          addedRowsCount += manifest.addedRowsCount();
     }
      writer.add(manifest);
}
```